internal/component/statefulset/statefulset.go

```diff
@@ -0,0 +1,217 @@
+	}
+	return r.createOrPatch(ctx, etcd)
+}
+
+// TriggerDelete triggers the deletion of the statefulset for the given Etcd.
+func (r _resource) TriggerDelete(ctx component.OperatorContext, etcdObjMeta metav1.ObjectMeta) error {
+	objectKey := getObjectKey(etcdObjMeta)
+	ctx.Logger.Info("Triggering deletion of StatefulSet", "objectKey", objectKey)
+	if err := r.client.Delete(ctx, emptyStatefulSet(etcdObjMeta)); err != nil {
+		if apierrors.IsNotFound(err) {
+			ctx.Logger.Info("No StatefulSet found, Deletion is a No-Op", "objectKey", objectKey.Name)
+			return nil
+		}
+		return druiderr.WrapError(err,
+			ErrDeleteStatefulSet,
+			"TriggerDelete",
+			fmt.Sprintf("Failed to delete StatefulSet: %v for etcd %v", objectKey, druidv1alpha1.GetNamespaceName(etcdObjMeta)))
+	}
+	ctx.Logger.Info("deleted", "component", "statefulset", "objectKey", objectKey)
+	return nil
+}
+
+// getExistingStatefulSet gets the existing statefulset if it exists.
+// If it is not found, it simply returns nil. Any other errors are returned as is.
+func (r _resource) getExistingStatefulSet(ctx component.OperatorContext, etcdObjMeta metav1.ObjectMeta) (*appsv1.StatefulSet, error) {
+	sts := emptyStatefulSet(etcdObjMeta)
+	if err := r.client.Get(ctx, getObjectKey(etcdObjMeta), sts); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+	return sts, nil
+}
+
+// createOrPatchWithReplicas ensures that the StatefulSet is updated with all changes from passed in etcd but the replicas set on the StatefulSet
+// are taken from the passed in replicas and not from the etcd component.
+func (r _resource) createOrPatchWithReplicas(ctx component.OperatorContext, etcd *druidv1alpha1.Etcd, replicas int32) error {
+	desiredStatefulSet := emptyStatefulSet(etcd.ObjectMeta)
+	mutatingFn := func() error {
+		if builder, err := newStsBuilder(r.client, ctx.Logger, etcd, replicas, r.useEtcdWrapper, r.imageVector, desiredStatefulSet); err != nil {
+			return druiderr.WrapError(err,
+				ErrSyncStatefulSet,
+				"Sync",
+				fmt.Sprintf("Error initializing StatefulSet builder for etcd %v", druidv1alpha1.GetNamespaceName(etcd.ObjectMeta)))
+		} else {
+			return builder.Build(ctx)
+		}
+	}
+	opResult, err := controllerutils.GetAndCreateOrStrategicMergePatch(ctx, r.client, desiredStatefulSet, mutatingFn)
+	if err != nil {
+		return druiderr.WrapError(err,
+			ErrSyncStatefulSet,
+			"Sync",
+			fmt.Sprintf("Error creating or patching StatefulSet: %s for etcd: %v", desiredStatefulSet.Name, druidv1alpha1.GetNamespaceName(etcd.ObjectMeta)))
+	}
+
+	ctx.Logger.Info("triggered creation of statefulSet", "statefulSet", getObjectKey(etcd.ObjectMeta), "operationResult", opResult)
+	return nil
+}
+
+// createOrPatch updates StatefulSet taking changes from passed in etcd component.
+func (r _resource) createOrPatch(ctx component.OperatorContext, etcd *druidv1alpha1.Etcd) error {
```

Comment:
Not really. if you check the logic for `handlePeerTLSChanges()`, you will see that we first patch the sts spec with new mount paths for peer TLS cert secrets, but we need to keep the replicas the same as before, especially required for the gardener case where peer TLS is enabled and replicas is changed from 1 to 3 in the same change. To handle this, we call `createOrPatchWithReplicas()` with old replicas value.

But once the peer TLS change has been handled (or not required to be handled), we go ahead with the full patch which uses the etcd spec.replicas value itself. And since a regular sts patch call anywhere in the code should not bother about this extra `replicas` field, we abstract it out using a generic `createOrPatch()` function.

Reply:
Okay, makes sense
Thanks for the explanation